### PR TITLE
GC: wasmparser: Add user defined struct types

### PR DESCRIPTION
**Pull request by imikushin**
Add user defined struct types to wasmparser/wasmprinter, with a roundtrip test case.

### CODE REVIEW

File: crates/wasmparser/src/validator/types.rs [hunk 1]
```diff
@@ -277,9 +287,11 @@ impl Type {
     }
 
     pub(crate) fn type_size(&self) -> u32 {
+        // TODO calculate actual size for func, array, struct (#1036)
```
Comment (fitzgen) [inline crates/wasmparser/src/validator/types.rs:290]: Can you file a follow up issue to track this?

```suggestion
        // TODO(#1234): calculate actual size for func, array, struct (#1036)
```

**Review by fitzgen — COMMENTED** [review]
Looking good! Some minor nitpicks below.

File: crates/wasmparser/src/validator/types.rs (hunk 1 above)
Comment (imikushin) [reply crates/wasmparser/src/validator/types.rs:290]: Is there anything that #1036 isn't covering?

File: crates/wasm-mutate/src/module.rs [hunk 2]
```diff
@@ -67,7 +67,10 @@ impl TryFrom<wasmparser::Type> for TypeInfo {
                     .collect(),
             })),
             wasmparser::Type::Array(_) => {
-                unimplemented!("Array and struct types are not supported yet.")
+                unimplemented!("Array types are not supported yet.")
+            }
+            wasmparser::Type::Struct(_) => {
+                unimplemented!("Struct types are not supported yet.")
```
[the remark this answers is not in the document]
Comment (imikushin) [reply crates/wasm-mutate/src/module.rs]: Done.

File: crates/wasm-mutate/src/mutators/translate.rs [hunk 3]
```diff
@@ -137,7 +137,8 @@ pub fn type_def(t: &mut dyn Translator, ty: Type, s: &mut TypeSection) -> Result
             );
             Ok(())
         }
-        Type::Array(_) => unimplemented!("Array and struct types are not supported yet."),
+        Type::Array(_) => unimplemented!("Array types are not supported yet."),
+        Type::Struct(_) => unimplemented!("Struct types are not supported yet."),
```
[the remark this answers is not in the document]
Comment (imikushin) [reply crates/wasm-mutate/src/mutators/translate.rs]: Done.

File: crates/wasmparser/src/limits.rs [hunk 4]
```diff
@@ -34,6 +34,7 @@ pub const MAX_WASM_TABLES: usize = 100;
 pub const MAX_WASM_MEMORIES: usize = 100;
 pub const MAX_WASM_TAGS: usize = 1_000_000;
 pub const MAX_WASM_BR_TABLE_SIZE: usize = MAX_WASM_FUNCTION_SIZE;
+pub const MAX_WASM_STRUCT_FIELDS: usize = 100_000;
```
[the remark this answers is not in the document]
Comment (imikushin) [reply crates/wasmparser/src/limits.rs]: Done.

**Review by fitzgen — APPROVED** [review]
Thanks!

**Pull request merged**